File: packages/py/analytics/src/bolsa_analytics/cognitive/mfe_mae.py
"""MFE/MAE excursion advisory (ADR-031 Ciclo 5.3).

Read-only metrics: peak favorable/adverse in R. No protect/exit tips, no fill gate.
"""

from __future__ import annotations

from typing import Any, Literal, Sequence

MfeMaeStatus = Literal["none", "observe", "favorable", "adverse"]
MfeMaeWhy = Literal[
    "peak_from_bars",
    "close_proxy",
    "mae_ge_1r",
    "mfe_ge_1_5r",
    "missing_inputs",
]

MFE_MAE_KEY = "mfeMae"


def map_mfe_mae(
    *,
    direction: str | None = None,
    entry: float | None = None,
    structural_stop: float | None = None,
    last_close: float | None = None,
    bars: Sequence[dict[str, Any] | Any] | None = None,
) -> dict[str, Any]:
    """Peak MFE/MAE in R from bars when available; else lastClose proxy."""
    if direction not in ("long", "short") or entry is None or structural_stop is None:
        return {
            "status": "none",
            "mfeR": None,
            "maeR": None,
            "currentR": None,
            "why": ["missing_inputs"],
        }

    try:
        e = float(entry)
        stop = float(structural_stop)
    except (TypeError, ValueError):
        return {
            "status": "none",
            "mfeR": None,
            "maeR": None,
            "currentR": None,
            "why": ["missing_inputs"],
        }

    r = abs(e - stop)
    if r <= 0:
        return {
            "status": "none",
            "mfeR": None,
            "maeR": None,
            "currentR": None,
            "why": ["missing_inputs"],
        }

    sign = 1.0 if direction == "long" else -1.0
    why: list[str] = []

    current_r: float | None = None
    if last_close is not None:
        try:
            close = float(last_close)
            current_r = round(((close - e) / r) * sign, 4)
        except (TypeError, ValueError):
            current_r = None

    used_bars = False
    peak_fav = 0.0
    peak_adv = 0.0

    if bars:
        for bar in bars:
            try:
                if isinstance(bar, dict):
                    high = float(bar["high"])
                    low = float(bar["low"])
                else:
                    high = float(getattr(bar, "high"))
                    low = float(getattr(bar, "low"))
            except (KeyError, TypeError, ValueError, AttributeError):
                continue
            used_bars = True
            if direction == "long":
                peak_fav = max(peak_fav, high - e)
                peak_adv = max(peak_adv, e - low)
            else:
                peak_fav = max(peak_fav, e - low)
                peak_adv = max(peak_adv, high - e)

    if used_bars:
        why.append("peak_from_bars")
        mfe_r = round(max(peak_fav, 0.0) / r, 4)
        mae_r = round(max(peak_adv, 0.0) / r, 4)
    elif current_r is not None:
        why.append("close_proxy")
        mfe_r = round(max(current_r, 0.0), 4)
        mae_r = round(max(-current_r, 0.0), 4)
    else:
        return {
            "status": "none",
            "mfeR": None,
            "maeR": None,
            "currentR": None,
            "why": ["missing_inputs"],
        }

    if mae_r >= 1.0:
        why.append("mae_ge_1r")
    if mfe_r >= 1.5:
        why.append("mfe_ge_1_5r")

    status: MfeMaeStatus = "observe"
    if mae_r >= 1.0:
        status = "adverse"
    elif mfe_r >= 1.5:
        status = "favorable"

    return {
        "status": status,
        "mfeR": mfe_r,
        "maeR": mae_r,
        "currentR": current_r,
        "why": why,
    }


def build_mfe_mae_dict(
    *,
    direction: str | None = None,
    entry: float | None = None,
    structural_stop: float | None = None,
    last_close: float | None = None,
    bars: Sequence[dict[str, Any] | Any] | None = None,
) -> dict[str, Any]:
    return map_mfe_mae(
        direction=direction,
        entry=entry,
        structural_stop=structural_stop,
        last_close=last_close,
        bars=bars,
    )

File: packages/shared/src/cognitive/mfe-mae.ts
/**
 * MFE/MAE excursion advisory (ADR-031 Ciclo 5.3).
 * Read-only metrics: peak favorable/adverse in R. No protect/exit tips, no fill gate.
 */

import type { TradePlanDirectionV1 } from "./trade-plan.js";

export type MfeMaeStatusV1 = "none" | "observe" | "favorable" | "adverse";

export type MfeMaeWhyV1 =
  | "peak_from_bars"
  | "close_proxy"
  | "mae_ge_1r"
  | "mfe_ge_1_5r"
  | "missing_inputs";

export type MfeMaeBarV1 = {
  high?: number | null;
  low?: number | null;
};

export type MfeMaeV1 = {
  status: MfeMaeStatusV1;
  mfeR: number | null;
  maeR: number | null;
  currentR: number | null;
  why: MfeMaeWhyV1[];
};

export type MapMfeMaeInput = {
  direction?: TradePlanDirectionV1 | null;
  entry?: number | null;
  structuralStop?: number | null;
  lastClose?: number | null;
  bars?: readonly MfeMaeBarV1[] | null;
};

function finite(n: unknown): n is number {
  return typeof n === "number" && Number.isFinite(n);
}

function round4(n: number): number {
  return Math.round(n * 10000) / 10000;
}

/**
 * Peak MFE/MAE in R from bars when available; else lastClose proxy.
 * Does not emit protect/trail suggestions.
 */
export function mapMfeMae(input: MapMfeMaeInput): MfeMaeV1 {
  const direction = input.direction;
  const entry = input.entry;
  const stop = input.structuralStop;
  const lastClose = input.lastClose;

  if (
    (direction !== "long" && direction !== "short") ||
    !finite(entry) ||
    !finite(stop)
  ) {
    return {
      status: "none",
      mfeR: null,
      maeR: null,
      currentR: null,
      why: ["missing_inputs"],
    };
  }

  const R = Math.abs(entry - stop);
  if (R <= 0) {
    return {
      status: "none",
      mfeR: null,
      maeR: null,
      currentR: null,
      why: ["missing_inputs"],
    };
  }

  const sign = direction === "long" ? 1 : -1;
  const why: MfeMaeWhyV1[] = [];

  let currentR: number | null = null;
  if (finite(lastClose)) {
    currentR = round4(((lastClose - entry) / R) * sign);
  }

  const bars = input.bars;
  let usedBars = false;
  let peakFav = 0;
  let peakAdv = 0;

  if (Array.isArray(bars) && bars.length > 0) {
    for (const bar of bars) {
      const high = bar?.high;
      const low = bar?.low;
      if (!finite(high) || !finite(low)) continue;
      usedBars = true;
      if (direction === "long") {
        peakFav = Math.max(peakFav, high - entry);
        peakAdv = Math.max(peakAdv, entry - low);
      } else {
        peakFav = Math.max(peakFav, entry - low);
        peakAdv = Math.max(peakAdv, high - entry);
      }
    }
  }

  let mfeR: number;
  let maeR: number;

  if (usedBars) {
    why.push("peak_from_bars");
    mfeR = round4(Math.max(peakFav, 0) / R);
    maeR = round4(Math.max(peakAdv, 0) / R);
  } else if (currentR != null) {
    why.push("close_proxy");
    mfeR = round4(Math.max(currentR, 0));
    maeR = round4(Math.max(-currentR, 0));
  } else {
    return {
      status: "none",
      mfeR: null,
      maeR: null,
      currentR: null,
      why: ["missing_inputs"],
    };
  }

  if (maeR >= 1) why.push("mae_ge_1r");
  if (mfeR >= 1.5) why.push("mfe_ge_1_5r");

  let status: MfeMaeStatusV1 = "observe";
  if (maeR >= 1) status = "adverse";
  else if (mfeR >= 1.5) status = "favorable";

  return {
    status,
    mfeR,
    maeR,
    currentR,
    why,
  };
}
